internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,111 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the Token method.
+type Tokener interface {
+	Token(context.Context) (string, error)
+}
+
+// LazyToken is a struct that handles token acquisition and caching.
+type LazyToken struct {
+	// Url represents the URL used for acquiring the token.
+	Url string
+	// ClientId is the client ID used for authentication.
+	ClientId string
+	// ClientSecret is the client secret used for authentication.
+	ClientSecret string
+	// token holds the currently cached token.
+	token string
+	// Expiration stores the expiration time of the current token.
+	Expiration time.Time
+	// Mutex ensures safe concurrent access to the token and expiration fields.
+	Mutex sync.Mutex
+}
+type tokenResponse struct {
+	AccessToken string    `json:"access_token"`
+	Expiration  time.Time `json:"expiration"`
+}
+
+// Token returns a valid token, acquiring a new one if necessary.
+func (lt *LazyToken) Token(ctx context.Context) (string, error) {
+	lt.Mutex.Lock()
+	defer lt.Mutex.Unlock()
+
+	if lt.token == "" || time.Now().After(lt.Expiration.Add(-time.Minute)) {
+		// Acquire a new token and set the expiration time
+		tokenRes, err := lt.RequestToken()
+		if err != nil {
+			return "", err
+		}
+
+		lt.token = tokenRes.AccessToken
+		lt.Expiration = tokenRes.Expiration
+		logrus.WithContext(ctx).Infof("Acquired new token %s with expiration at: %s", lt.token, lt.Expiration)
+	} else {
+		logrus.WithContext(ctx).Infof("Token requested: %s (expires at %s)", lt.token, lt.Expiration)
+	}
+
+	return lt.token, nil
+}
+
+// RequestToken is a placeholder function that simulates token acquisition.
+func (lt *LazyToken) RequestToken() (*tokenResponse, error) {
+	req, err := http.NewRequest("POST", lt.Url, nil)
+	if err != nil {
+		return nil, err
+
+	}
+	data := url.Values{}
+
+	if time.Now().After(lt.Expiration) {
+		data.Set("grant_type", "refresh_token")
+		data.Set("client_id", lt.ClientId)
+	} else if lt.ClientSecret != "" {
+		data.Set("grant_type", "client_credentials")
+		data.Set("client_id", lt.ClientId)
+		data.Set("client_secret", lt.ClientSecret)
+	}
```

Comment:
This does not appear to be correct:

* If client id is not empty string AND (token is not empty string or token has expired) then do "refresh_token" call
* Else if client id and secret are both not empty string, do "client_credentials" calls.
* Else return error that client id, secret or token must be set.